File: Lec5/L5.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

#define maxV 100
int V, E;
bool visited[maxV]; // visited Vs
int path[maxV]; // previous V of this V
vector<int> graph[maxV]; // adjacent Vs of this V

void BFS(int s) {
    for (int i=0; i<V; ++i) {
        visited[i] = 0;
        path[i] = -1;
    }

    queue<int> q;
    visited[s] = 1;
    q.push(s);
    
    while(!q.empty()) {
        int v = q.front();
        q.pop();
        for (int i=0; i<graph[v].size(); ++i) {
            int adja = graph[v][i];
            if (!visited[adja]) {
                visited[adja] = 1;
                q.push(adja);
                path[adja] = v;
            }
        }
    }
}

void printPath(int s, int f) {
    int ans[maxV];
    int i = 0;
    if (f == s) {
        cout << s;
        return;
    }

    if (path[f] == -1) {
        cout << "No path\n";
        return;
    }

    while (true) {
        ans[i++] = f;
        f = path[f];
        if (f == s) {
            ans[i++] = s;
            break;
        }
    }

    while (i--) {
        cout << ans[i] << ' ';
    }
}

void printPathRec(int s, int f) {
    if (s == f) cout << f;
    else {
        if (path[f] == -1) cout << "No path\n";
        else {
            printPathRec(s, path[f]);
            cout << f << ' ';
        }
    }
}

void readAdjacencyMatrix() {
    int mark;
    for (int i=0; i<V; i++) {
        for (int j=0; j<V; j++) {
            cin >> mark;
            if (mark == 1) {
                graph[i].push_back(j);
            }
        }
    }
}

void readEdgeList() {
    // read edge list into graph
    int u, v;
    for (int i=0; i<E; ++i) {
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
}

int main() {
    cin >> V >> E;
    readAdjacencyMatrix();
    readEdgeList();

    int s = 0;
    int f = 5;
    BFS(s);
    printPath(s, f);
    return 0;
}

File: Lec5/L5.py
from queue import Queue
maxV = 100
V = E = None
visited = [False]*maxV
path = [-1]*maxV
graph = [[] for _ in range(maxV)]

def BFS(s):
    for i in range(V):
        visited[i] = False
        path[i] = -1
    q = Queue()
    visited[s] = True
    q.put(s)
    while not q.empty():
        v = q.get()
        for adja in graph[v]:
            if not visited[adja]:
                visited[adja] = True
                q.put(adja)
                path[adja] = v


def printPath(s, f):
    ans = []
    if f == s:
        print(s)
        return
    if path[f] == -1:
        print("No path")
        return
    while True:
        ans.append(f)
        f = path[f]
        if f == s:
            ans.append(f)
            break
    
    for i in range(len(ans)-1, -1, -1):
        print(ans[i], end=' ')

def printPathRec(s, f):
    if s == f:
        print(f, end=' ')
    else:
        if path[f] == -1:
            print("No path")
        else:
            printPathRec(s, path[f])
            print(f, end=' ')
    
if __name__ == '__main__':
    V, E = map(int, input().split())
    for i in range(E):
        u, v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)
    s = 0
    f = 5
    BFS(s)
    printPath(s, f)